### DISCUSSION TITLE
How to change CLIP Skip?

### DISCUSSION THREAD

**Discussion by ToniMagi** [lllyasviel/Fooocus#2170] [Q&A] [answered]
I would like to know how to configure the CLIP Skip. I need to set it to 2 and by default I think it is -2. There are some models that do not generate properly if it does not have a value of 2. Please, some advice or guide? Thanks!

**Comment by gltchcom** [comment]
I have heard SDXL ignores CLIP Skip. It seems to be true in my testing, but I'm new to all this. I just don't know why so many SDXL model creators are telling people to change CLIP Skip if that isn't a thing in SDXL. I hope someone can clarify.

**Reply by ToniMagi** [reply]
I come from AUTO1111 and it would be ideal if, if necessary, they implemented the possibility of changing this parameter in a way as simple as in AUTO1111. Some SDXL models even their developers indicate that it does not generate images well if you do not change the CLIP Skip to 2. So I suppose it is necessary in SDXL as well. I don't know, I hope that someone who understands the subject can help with a solution or that it will be implemented in Fooocus in future versions. Thanks.

**Comment by AkoZ** [comment]
_As i'm also confused with that clip skip, i try a copilot explantions, which were a bit outdated at first.
It may be useful to create an encyclopedia of the pramaters one day ..  in a same useful dico_
Of course, feel free to complete or may be correct those answers.

Clip Skip is a feature in Stable Diffusion that allows you to control the level of detail in the images you generate. [It does this by skipping some layers of the CLIP model during the image generation process](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674)[1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674)[2](https://medium.com/@novita.ai/understanding-clip-skip-stable-diffusion-417b45f20645).

The CLIP model is a powerful tool developed by OpenAI that can understand the relationship between images and text. [It transforms your text prompt into a numerical representation that guides the image generation](https://github.com/openai/CLIP)[3](https://github.com/openai/CLIP)[4](https://openai.com/research/clip/).

[The Clip Skip value represents the number of layers to skip from the bottom of the CLIP model’s architecture](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674)[1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674).
Clip Skip is a feature in Stable Diffusion that allows you to control the level of detail in the images you generate. It does this by skipping some layers of the CLIP model during the image generation process.

The CLIP model is a powerful tool developed by OpenAI that can understand the relationship between images and text. It transforms your text prompt into a numerical representation that guides the image generation. This transformation process involves several layers, each adding more detail and complexity to the representation.

The Clip Skip value represents the number of layers to skip from the bottom of the CLIP model’s architecture.
Here’s how different Clip Skip values affect the image generation:

Clip Skip = 0: The algorithm uses all layers of the CLIP model. [This gives you the most detailed image, but it’s also the most computationally intensive](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674)[1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674).
Clip Skip = 1: The algorithm skips the last layer of the CLIP model. [This gives you a slightly less detailed image, but it’s less computationally intensive](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674)[1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674).
Clip Skip = 2: The algorithm skips the last two layers of the CLIP model. [This means that in an SD1.x model where the CLIP has 12 layers, you would stop at the 10th layer](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674)[1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674).
[Remember, the optimal balance between speed and quality can vary depending on your specific use case and the model you’re using](https://medium.com/@novita.ai/understanding-clip-skip-stable-diffusion-417b45f20645)[2](https://medium.com/@novita.ai/understanding-clip-skip-stable-diffusion-417b45f20645).

The CLIP model is a neural network trained on a variety of (image, text) pairs. [It can be instructed in natural language to predict the most relevant text snippet, given an image, without directly optimizing for the task, similarly to the zero-shot capabilities of GPT-2 and 3](https://github.com/openai/CLIP)[3](https://github.com/openai/CLIP)[4](https://openai.com/research/clip/).

More detailed informations refer to these resources:

[What exactly is clip skip?](https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/5674)
[Understanding Clip Skip Stable Diffusion](https://medium.com/@novita.ai/understanding-clip-skip-stable-diffusion-417b45f20645)
[Stable Diffusion Skip Clip](https://nightcafe.studio/blogs/info/stable-diffusion-skip-clip/)
[What is Clip Skip in Stable Diffusion and What Does It Do?](https://ambcrypto.com/blog/what-is-clip-skip-in-stable-diffusion-and-what-does-it-do/)
[GitHub - openai/CLIP: CLIP (Contrastive Language-Image Pretraining)](https://github.com/openai/CLIP)
[CLIP: Connecting text and images - OpenAI](https://openai.com/research/clip/)
[CLIP - Hugging Face](https://huggingface.co/docs/transformers/model_doc/clip)
[New CLIP model aims to make Stable Diffusion even better - THE DECODER](https://the-decoder.com/new-clip-model-aims-to-make-stable-diffusion-even-better/)
[What is OpenAI’s CLIP and how to use it? - Roboflow Blog](https://blog.roboflow.com/openai-clip/)

**Comment by funx24x7** [comment]
Try turning off any styles to see if that helps. The first time I tried a model that suggested clip skip 2 I was getting random blobs in Fooocus while it worked in AUTO1111. However after some more messing around it seems that the cause was too large/complex of a prompt, the "Fooocus Enhance" style in particular killed the image for me every time.

Some models come with a recommended sampler, you can access these under the "Developer Debug Mode" box in the Advanced tab.

Lastly, I found a fork of Fooocus that includes clip skip ([Fooocus-MRE](https://github.com/MoonRide303/Fooocus-MRE)) but for some reason you can't set it higher than -1 without going into the config files. Also it downloads like 17gb of models so heads up there. Hope this helps.

**Reply by mashb1t** [reply, marked as answer]
in theory it should be sufficient to apply https://github.com/MoonRide303/Fooocus-MRE/blob/e4415d1d983ffaaeeaacdc23489a335c078641fe/modules/default_pipeline.py#L197-L203 and set/load it accordingly, but has to be tested.

**Reply by BNP1111** [reply]
Hello, if I modify the code according to the link you posted, do I still need to add like 
"clip_skips": {
    "base_clip_skip": 2,
    "refiner_clip_skip": 2
}
in config.txt?

**Reply by mashb1t** [reply]
@BNP1111 only if you use the config values in the code you've adjusted

**Comment by ToniMagi** [comment]
Thank you all for the suggestions, but I already found the solution: I have tried AUTO1111 Forge and it is spectacular. VRAM management is optimized and allows the same parameters as AUTO1111 to be modified (obviously including CLIP Skip). Try it, it's wonderful and goes like silk! Closed.

**Comment by mashb1t** [comment]
CLIP Skip has been implemented in https://github.com/lllyasviel/Fooocus/pull/2999 and is part of the upcoming release 2.4.0.
